File: ql/tos2ql.py
#!/usr/bin/env python3
"""
tos2ql.py — Convert Atari ST TOS executable to self-relocating QDOS binary.

TOS format:
  28-byte header: magic, text_size, data_size, bss_size, sym_size, ...
  text section (code)
  data section
  symbol table (skip)
  relocation table: first longword = offset of first fixup,
    then bytes: 0=end, 1=add 254, 2-255=delta to next fixup

Output QDOS format:
  Q-emulator header (30 bytes)
  Self-relocating stub (moves code, applies fixups, calls main)
  Code + data
  Relocation table (compact: list of 16-bit deltas, 0=end)
"""
import struct
import sys

def read_tos(filename):
    with open(filename, 'rb') as f:
        data = f.read()

    # TOS header (28 bytes)
    magic = struct.unpack_from('>H', data, 0)[0]
    text_size = struct.unpack_from('>I', data, 2)[0]
    data_size = struct.unpack_from('>I', data, 6)[0]
    bss_size = struct.unpack_from('>I', data, 10)[0]
    sym_size = struct.unpack_from('>I', data, 14)[0]

    print(f"TOS: text={text_size} data={data_size} bss={bss_size} sym={sym_size}")

    text_start = 28
    data_start = text_start + text_size
    sym_start = data_start + data_size
    reloc_start = sym_start + sym_size

    code_data = data[text_start:data_start + data_size]

    # Parse relocation table
    fixups = []
    pos = reloc_start
    if pos < len(data):
        first = struct.unpack_from('>I', data, pos)[0]
        pos += 4
        if first != 0:
            offset = first
            fixups.append(offset)
            while pos < len(data):
                b = data[pos]
                pos += 1
                if b == 0:
                    break
                elif b == 1:
                    offset += 254
                else:
                    offset += b
                    fixups.append(offset)

    print(f"Relocations: {len(fixups)} fixups")
    return code_data, fixups, bss_size


def make_qdos_binary(code_data, fixups, bss_size, outfile):
    # Self-relocating stub in 68K machine code
    # This runs at offset 0 of the loaded job.
    # It computes load_base, walks the relocation table, patches addresses,
    # then jumps to the original entry point.
    #
    # Register usage:
    #   a0 = current fixup pointer in code
    #   a1 = relocation table pointer
    #   a2 = load base (actual address of code start)
    #   d0 = current offset
    #   d1 = delta value

    # The stub needs to:
    # 1. lea stub_start(pc), a2  -> a2 = actual runtime address of stub
    # 2. lea code_start(pc), a3  -> a3 = actual runtime address of code
    # 3. For each fixup: read 32-bit value at code+offset, add a3, write back
    # 4. jmp to code entry point

    stub_code = bytearray()

    # Offsets we'll patch later
    # lea _stub(pc), a2         ; 4 bytes
    # lea _code(pc), a3         ; 4 bytes  (code starts after stub + reloc table)
    # lea _reloc(pc), a1        ; 4 bytes

    code_offset = None  # will compute after we know stub+reloc size

    # Build relocation table as 16-bit deltas (more compact than TOS format)
    reloc_table = bytearray()
    prev = 0
    for fixup in fixups:
        delta = fixup - prev
        while delta > 65534:
            reloc_table += struct.pack('>H', 65534)
            reloc_table += struct.pack('>H', 0)  # 0 = skip (no fixup, just advance)
            delta -= 65534
        reloc_table += struct.pack('>H', delta)
        prev = fixup
    reloc_table += struct.pack('>H', 0)  # end marker
    reloc_table += struct.pack('>H', 0)  # padding for alignment

    # Compute sizes
    # Stub is small fixed-size code (~40 bytes)
    # Then relocation table
    # Then code+data

    # Build the stub assembly as machine code bytes:
    #
    # _stub:
    #   lea     _code(pc),a3          ; 41FB xxxx  (PC-relative to code)
    #   lea     _reloc(pc),a1         ; 43FA xxxx  (PC-relative to reloc table)
    #   move.l  a3,a2                 ; 2448       (a2 = code base for fixups)
    # .loop:
    #   move.w  (a1)+,d0              ; 3019       (read 16-bit delta)
    #   beq.s   .done                 ; 6700 xxxx  (0 = end)
    #   add.l   d0,a0                 ; D1C0... wait, we need a0
    #
    # Actually let me just hardcode the stub bytes.

    # Simpler stub approach:
    # We know the exact layout:
    #   [stub code] [reloc table] [code+data]
    # stub code lea's are PC-relative to known offsets.

    stub_size = 36  # we'll make it exactly this
    reloc_offset = stub_size  # reloc table starts right after stub
    code_start_offset = stub_size + len(reloc_table)

    # Ensure code starts at even address
    if code_start_offset & 1:
        reloc_table += b'\x00'
        code_start_offset += 1

    # Now build stub machine code:
    s = bytearray()

    # lea code(pc), a3  -->  47FA disp16
    disp_to_code = code_start_offset - (len(s) + 2)  # PC at instruction + 2
    s += bytes([0x47, 0xFA]) + struct.pack('>h', disp_to_code)

    # lea reloc(pc), a1  -->  43FA disp16
    disp_to_reloc = reloc_offset - (len(s) + 2)
    s += bytes([0x43, 0xFA]) + struct.pack('>h', disp_to_reloc)

    # movea.l a3, a0  -->  2048 + 0BC0... no
    # move.l a3, a0  -->  2048... wait
    # movea.l a3,a0 = 204B
    s += bytes([0x20, 0x4B])  # movea.l a3, a0  (a0 = code base, used as current ptr)

    # Relocation loop:
    # .loop:
    loop_offset = len(s)
    #   move.w (a1)+, d0  -->  3019
    s += bytes([0x30, 0x19])
    #   beq.s .done  -->  6700 xxxx (branch if zero)
    done_branch_pos = len(s)
    s += bytes([0x67, 0x00, 0x00, 0x00])  # placeholder, will patch

    #   ext.l d0  -->  48C0 (sign-extend word to long... but deltas are unsigned)
    #   Actually we want unsigned extend: andi.l #$FFFF, d0  or swap+clr+swap
    #   Simpler: moveq #0,d1; move.w d0,d1; use d1
    #   Or just: and.l #$FFFF, d0
    s += bytes([0x02, 0x80, 0x00, 0x00, 0xFF, 0xFF])  # andi.l #$FFFF, d0

    #   adda.l d0, a0  -->  D1C0
    s += bytes([0xD1, 0xC0])

    #   add.l a3, (a0)  -->  D793  (add.l a3, (a0))
    #   Wait: add.l An,(Am) is not valid. Need: move.l (a0),d1; add.l a3,d1; move.l d1,(a0)
    s += bytes([0x22, 0x10])  # move.l (a0), d1
    s += bytes([0xD2, 0x8B])  # add.l a3, d1
    s += bytes([0x20, 0x81])  # move.l d1, (a0)

    #   move.l a3, a0  -->  204B (reset a0 to code base for next delta)
    #   Wait no! The deltas are cumulative from the start. a0 should track current position.
    #   Actually, the deltas in our table are from the PREVIOUS fixup, not from start.
    #   So a0 advances by delta each time. Let me redo:

    # Oops, the approach is wrong. Let me restart the stub logic.
    # a0 starts at code_base.
    # Each delta: a0 += delta, then patch the long at (a0).
    # So we DON'T reset a0.

    # Let me rebuild properly:
    s = bytearray()

    # lea code(pc), a0  -->  41FA disp16
    disp_to_code = code_start_offset - (len(s) + 2)
    s += bytes([0x41, 0xFA]) + struct.pack('>h', disp_to_code)

    # lea reloc(pc), a1  -->  43FA disp16
    disp_to_reloc = reloc_offset - (len(s) + 2)
    s += bytes([0x43, 0xFA]) + struct.pack('>h', disp_to_reloc)

    # move.l a0, d2  -->  2408  (save code base in d2 for adding to fixups)
    s += bytes([0x24, 0x08])

    # .loop:
    loop_offset = len(s)
    # move.w (a1)+, d0  -->  3019
    s += bytes([0x30, 0x19])
    # beq.s .done
    done_branch_pos = len(s)
    s += bytes([0x67, 0x00])  # beq.s placeholder (8-bit displacement)

    # andi.l #$FFFF, d0  (unsigned extend 16->32)
    s += bytes([0x02, 0x80, 0x00, 0x00, 0xFF, 0xFF])

    # adda.l d0, a0  -->  D1C0
    s += bytes([0xD1, 0xC0])

    # add.l d2, (a0)  -->  D590
    s += bytes([0xD5, 0x90])

    # bra.s .loop
    loop_disp = loop_offset - (len(s) + 2)
    s += bytes([0x60, loop_disp & 0xFF])

    # .done:
    done_offset = len(s)
    # Patch the beq.s
    beq_disp = done_offset - (done_branch_pos + 2)
    s[done_branch_pos + 1] = beq_disp & 0xFF

    # jmp (a0) ... no, jump to entry point
    # The TOS entry point is at offset 0 of the text section.
    # a0 is currently pointing somewhere in the code (last fixup location).
    # We need to jump to code_base.
    # move.l d2, a0  -->  2042
    s += bytes([0x20, 0x42])
    # jmp (a0)  -->  4ED0
    s += bytes([0x4E, 0xD0])

    # Pad to stub_size
    while len(s) < stub_size:
        s += bytes([0x4E, 0x71])  # nop

    if len(s) != stub_size:
        print(f"Warning: stub is {len(s)} bytes, expected {stub_size}")
        stub_size = len(s)
        # Recalculate offsets
        reloc_offset = stub_size
        code_start_offset = stub_size + len(reloc_table)
        if code_start_offset & 1:
            reloc_table += b'\x00'
            code_start_offset += 1
        # Re-patch the lea displacements
        disp_to_code = code_start_offset - 2
        struct.pack_into('>h', s, 2, disp_to_code)
        disp_to_reloc = reloc_offset - 6
        struct.pack_into('>h', s, 6, disp_to_reloc)

    # Build final binary
    binary = bytes(s) + bytes(reloc_table) + code_data

    # Prepend Q-emulator header
    hdr = b']!QDOS File Header'
    hdr += b'\x00'       # reserved
    hdr += bytes([15])   # wordlen (15 words = 30 bytes)
    hdr += bytes([0])    # access
    hdr += bytes([1])    # type = executable
    hdr += struct.pack('>I', bss_size + 4096)  # data space (BSS + stack)
    hdr += struct.pack('>I', 0)   # reserved
    hdr += b'\x00' * (30 - len(hdr))

    # XTcc trailer (8 bytes at end) — for emulators that check file tail
    # instead of Q-emulator header (e.g., iQL)
    xtcc = b'XTcc' + struct.pack('>I', bss_size + 4096)

    with open(outfile, 'wb') as f:
        f.write(hdr)
        f.write(binary)
        f.write(xtcc)

    total = len(hdr) + len(binary) + len(xtcc)
    print(f"Output: {outfile} ({total} bytes)")
    print(f"  Stub: {stub_size}, Reloc: {len(reloc_table)}, Code: {len(code_data)}")
    print(f"  Headers: Q-emulator (start) + XTcc (tail)")


if __name__ == '__main__':
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} input.tos output_ql")
        sys.exit(1)
    code_data, fixups, bss_size = read_tos(sys.argv[1])
    make_qdos_binary(code_data, fixups, bss_size, sys.argv[2])

File: ql/tools/tos2ql/tos2ql.c
/*
 * tos2ql.c — Convert Atari ST TOS executable to self-relocating QDOS binary.
 *
 * TOS format:
 *   28-byte header: magic, text_size, data_size, bss_size, sym_size, ...
 *   text section (code)
 *   data section
 *   symbol table (skip)
 *   relocation table: first longword = offset of first fixup,
 *     then bytes: 0=end, 1=add 254, 2-255=delta to next fixup
 *
 * Output QDOS format:
 *   Self-relocating stub (36 bytes)
 *   Relocation table (compact: 16-bit deltas, 0=end)
 *   Code + data
 *   XTcc trailer (8 bytes)
 *
 * Usage: tos2ql input.tos output_ql
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

/* Read big-endian values from buffer */
static uint16_t rd16(const uint8_t *p) { return (p[0] << 8) | p[1]; }
static uint32_t rd32(const uint8_t *p) { return (p[0] << 24) | (p[1] << 16) | (p[2] << 8) | p[3]; }

/* Write big-endian values to buffer */
static void wr16(uint8_t *p, uint16_t v) { p[0] = v >> 8; p[1] = v & 0xFF; }
static void wr32(uint8_t *p, uint32_t v) {
    p[0] = (v >> 24) & 0xFF; p[1] = (v >> 16) & 0xFF;
    p[2] = (v >> 8) & 0xFF; p[3] = v & 0xFF;
}

#define MAX_FIXUPS 8192
#define STUB_SIZE 36

int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s input.tos output_ql\n", argv[0]);
        return 1;
    }

    /* Read input file */
    FILE *f = fopen(argv[1], "rb");
    if (!f) { fprintf(stderr, "Cannot open %s\n", argv[1]); return 1; }
    fseek(f, 0, SEEK_END);
    long file_size = ftell(f);
    fseek(f, 0, SEEK_SET);
    uint8_t *data = malloc(file_size);
    fread(data, 1, file_size, f);
    fclose(f);

    /* Parse TOS header (28 bytes) */
    uint32_t text_size = rd32(data + 2);
    uint32_t data_size = rd32(data + 6);
    uint32_t bss_size  = rd32(data + 10);
    uint32_t sym_size  = rd32(data + 14);

    printf("TOS: text=%u data=%u bss=%u sym=%u\n", text_size, data_size, bss_size, sym_size);

    uint32_t text_start = 28;
    uint32_t data_start = text_start + text_size;
    uint32_t sym_start  = data_start + data_size;
    uint32_t reloc_start = sym_start + sym_size;
    uint32_t code_len = text_size + data_size;

    /* Parse relocation table */
    uint32_t fixups[MAX_FIXUPS];
    int fixup_count = 0;
    uint32_t pos = reloc_start;

    if (pos + 4 <= (uint32_t)file_size) {
        uint32_t first = rd32(data + pos);
        pos += 4;
        if (first != 0) {
            uint32_t offset = first;
            fixups[fixup_count++] = offset;
            while (pos < (uint32_t)file_size && fixup_count < MAX_FIXUPS) {
                uint8_t b = data[pos++];
                if (b == 0) break;
                else if (b == 1) offset += 254;
                else { offset += b; fixups[fixup_count++] = offset; }
            }
        }
    }

    printf("Relocations: %d fixups\n", fixup_count);

    /* Build compact relocation table (16-bit deltas) */
    int reloc_cap = (fixup_count + 2) * 4; /* worst case */
    uint8_t *reloc_table = calloc(reloc_cap, 1);
    int reloc_len = 0;
    uint32_t prev = 0;

    for (int i = 0; i < fixup_count; i++) {
        uint32_t delta = fixups[i] - prev;
        while (delta > 65534) {
            wr16(reloc_table + reloc_len, 65534); reloc_len += 2;
            wr16(reloc_table + reloc_len, 0);     reloc_len += 2; /* skip marker */
            delta -= 65534;
        }
        wr16(reloc_table + reloc_len, (uint16_t)delta); reloc_len += 2;
        prev = fixups[i];
    }
    wr16(reloc_table + reloc_len, 0); reloc_len += 2; /* end marker */
    wr16(reloc_table + reloc_len, 0); reloc_len += 2; /* alignment padding */

    /* Compute offsets */
    int reloc_offset = STUB_SIZE;
    int code_start_offset = STUB_SIZE + reloc_len;
    if (code_start_offset & 1) {
        reloc_table[reloc_len++] = 0; /* pad for alignment */
        code_start_offset++;
    }

    /* Build self-relocating 68K stub */
    uint8_t stub[STUB_SIZE];
    memset(stub, 0, STUB_SIZE);
    int sp = 0;

    /* lea code(pc), a0  -->  41FA disp16 */
    int disp_to_code = code_start_offset - (sp + 2);
    stub[sp++] = 0x41; stub[sp++] = 0xFA;
    wr16(stub + sp, (uint16_t)disp_to_code); sp += 2;

    /* lea reloc(pc), a1  -->  43FA disp16 */
    int disp_to_reloc = reloc_offset - (sp + 2);
    stub[sp++] = 0x43; stub[sp++] = 0xFA;
    wr16(stub + sp, (uint16_t)disp_to_reloc); sp += 2;

    /* move.l a0, d2  -->  2408  (save code base) */
    stub[sp++] = 0x24; stub[sp++] = 0x08;

    /* .loop: */
    int loop_offset = sp;

    /* move.w (a1)+, d0  -->  3019 */
    stub[sp++] = 0x30; stub[sp++] = 0x19;

    /* beq.s .done  -->  67 xx */
    int done_branch_pos = sp;
    stub[sp++] = 0x67; stub[sp++] = 0x00; /* placeholder */

    /* andi.l #$FFFF, d0  -->  0280 0000 FFFF */
    stub[sp++] = 0x02; stub[sp++] = 0x80;
    stub[sp++] = 0x00; stub[sp++] = 0x00;
    stub[sp++] = 0xFF; stub[sp++] = 0xFF;

    /* adda.l d0, a0  -->  D1C0 */
    stub[sp++] = 0xD1; stub[sp++] = 0xC0;

    /* add.l d2, (a0)  -->  D590 */
    stub[sp++] = 0xD5; stub[sp++] = 0x90;

    /* bra.s .loop */
    int loop_disp = loop_offset - (sp + 2);
    stub[sp++] = 0x60; stub[sp++] = (uint8_t)(loop_disp & 0xFF);

    /* .done: */
    int done_offset = sp;
    stub[done_branch_pos + 1] = (uint8_t)((done_offset - (done_branch_pos + 2)) & 0xFF);

    /* move.l d2, a0  -->  2042 */
    stub[sp++] = 0x20; stub[sp++] = 0x42;

    /* jmp (a0)  -->  4ED0 */
    stub[sp++] = 0x4E; stub[sp++] = 0xD0;

    /* Pad with nops */
    while (sp < STUB_SIZE) {
        stub[sp++] = 0x4E; stub[sp++] = 0x71; /* nop */
    }

    /* Write output */
    f = fopen(argv[2], "wb");
    if (!f) { fprintf(stderr, "Cannot create %s\n", argv[2]); free(data); free(reloc_table); return 1; }

    fwrite(stub, 1, STUB_SIZE, f);
    fwrite(reloc_table, 1, reloc_len, f);
    fwrite(data + text_start, 1, code_len, f);

    /* XTcc trailer (8 bytes) — marks file as executable */
    uint8_t xtcc[8];
    memcpy(xtcc, "XTcc", 4);
    wr32(xtcc + 4, bss_size + 4096);
    fwrite(xtcc, 1, 8, f);

    long total = STUB_SIZE + reloc_len + code_len + 8;
    fclose(f);

    printf("Output: %s (%ld bytes)\n", argv[2], total);
    printf("  Stub: %d, Reloc: %d, Code: %u\n", STUB_SIZE, reloc_len, code_len);
    printf("  Header: XTcc trailer (%u data space)\n", bss_size + 4096);

    free(data);
    free(reloc_table);
    return 0;
}
